### PR TITLE
Allow selection of specific BMH per compute node in OSBMS

### PR DESCRIPTION
**Pull request by abays**
Adds an optional `bmhLabelSelector` field in the instanceSpec `OpenStackBaremetalSet` so that a specific BMH can be chosen per instance if desired.

jira: [OSPRH-10282](https://issues.redhat.com//browse/OSPRH-10282)
Co-Authored-By: @rabi ramishra@redhat.com

### CODE REVIEW

**Comment by abays** [comment]
recheck

File: api/v1beta1/openstackbaremetalset.go [hunk 1]
```diff
@@ -90,10 +90,10 @@ func VerifyBaremetalSetScaleUp(
 	l logr.Logger,
 	instance *OpenStackBaremetalSet,
 	allBmhs *metal3v1.BareMetalHostList,
-	existingBmhs *metal3v1.BareMetalHostList) ([]string, error) {
+	existingBmhs *metal3v1.BareMetalHostList) ([]metal3v1.BareMetalHost, error) {
 	// How many new BaremetalHost allocations do we need (if any)?
```
Comment (rabi) [inline api/v1beta1/openstackbaremetalset.go:94]: This function uses `instance.Spec.BmhLabelSelector` when checking for scale-up. I guess we also have to match the label in the instancespec if there is one?

**Review by rabi — COMMENTED** [review]
What do you think about below?

1. First match the BMH specific label in instanceSpec
2.  As a fallback match label selector for baremetalSet? This could be irrelevant in cases when someone has a label for BMH in instanceSpec

We can match both if provided though. However, we have to ensure that scale-out logic also includes that.

File: api/v1beta1/openstackbaremetalset.go (hunk 1 above)
Comment (abays) [reply api/v1beta1/openstackbaremetalset.go:94]: Fair point.  We should fail early in the webhook for the new individual selector label just like we do for the existing group selector label.

File: controllers/openstackbaremetalset_controller.go [hunk 2]
```diff
@@ -164,6 +164,7 @@ func (r *OpenStackBaremetalSetReconciler) SetupWithManager(mgr ctrl.Manager) err
 	openshiftMachineAPIBareMetalHostsFn := handler.EnqueueRequestsFromMapFunc(func(_ context.Context, o client.Object) []reconcile.Request {
 		result := []reconcile.Request{}
 		label := o.GetLabels()
+		r.Log.Info(fmt.Sprintf("#####BareMetalHost object triggering reconcile %s", o.GetName()))
```
Comment (abays) [inline controllers/openstackbaremetalset_controller.go:167]: Is this leftover debug or did we want this going forward?

Comment (rabi) [reply controllers/openstackbaremetalset_controller.go:167]: I can remove it either completely or remove the prefix hashes. We've it in some controllers though as it helps when looking at controller logs. https://github.com/openstack-k8s-operators/openstack-operator/blob/main/controllers/dataplane/openstackdataplanenodeset_controller.go#L701

Comment (rabi) [reply controllers/openstackbaremetalset_controller.go]: Removed!

**Comment by abays** [comment]
@rabi I tested your changes in my local environment in a few different ways.  Everything seemed to work properly, so I am good with this.

**Review by rabi — APPROVED** [review]
/lgtm

**Comment by rabi** [comment]
Merging this as it has been tested by both Andrew and me and +2 from Andrew.

**Pull request merged**